include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -6906,6 +6940,14 @@ FEEvaluation<dim,
   for (unsigned int i = 0; i < VectorizedArrayType::size(); ++i)
     this->cell_ids[i] = cell_index * VectorizedArrayType::size() + i;
 
+  if ((this->matrix_free == nullptr &&
+       (this->mapped_geometry->get_fe_values().get_update_flags() |
+        update_quadrature_points)) ||
+      this->mapping_data->quadrature_point_offsets.empty() == false)
+    this->quadrature_points =
+      &this->mapping_data->quadrature_points
+         [this->mapping_data->quadrature_point_offsets[this->cell]];
```

Comment:
I am not particularly happy to pull this rather expensive statement into the `reinit` call that gets unconditionally executed every time we pass around here - of course from the logic it makes sense to have everything in the base class `FEEvaluationData`.

Reply:
Is the problem the if-statement? We need to move all mapping-related data somehow to the base class.

Reply:
I just move the if-statement from the place where the quadrature point locations were used.

Reply:
Do we see an alternative cheaper solution? If not, I suggest we stick with this one here and try to find a solution at some other occasion (and by try I mean to really try hard, because these functions are not negligible for low polynomial degrees and in 2D).

Reply:
And yes, it was only moved over, but the difference is that previously one only paid the price for the access on demand of a quadrature point, whereas it now pops up for every use. (There is a reason for why this and the Jacobian gradient were not originally in the base class, the associated cost.)

Reply:
Would it be better to do it like this:

https://github.com/dealii/dealii/blob/5b4409689cd5cc76194a6b47720ca98aec2d7a04/include/deal.II/matrix_free/fe_evaluation.h#L7042-L7043

Also, I am wondering if the first part of the if statement should not go into the reinit function that takes cell accessors? I looks wrong here? Would this improve the situation?

Reply:
@kronbichler The last commit should contain all fixes. If you are happy, I'll squash.